### ISSUE TITLE
[Bug] `catalog.json` includes all nodes even when no nodes are selected during `dbt docs generate`

### ISSUE THREAD

**Issue by dbeatty10** [dbt-labs/dbt-core#8794]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

When no nodes are selected during `dbt docs generate`, then `target/catalog.json` includes **all** nodes (rather than 0 nodes).

### Expected Behavior

Expected that `target/catalog.json` would contain **0** nodes rather than **all** nodes.

i.e., expected that it would be the same as when running `dbt docs generate --empty-catalog`:
```json
{
  "metadata": {
    "dbt_schema_version": "https://schemas.getdbt.com/dbt/catalog/v1.json",
    "dbt_version": "1.7.0b2",
    "generated_at": "2023-10-07T13:34:04.627535Z",
    "invocation_id": "c09b4da4-35c5-418d-8a4b-447d2925f075",
    "env": {}
  },
  "nodes": {},
  "sources": {},
  "errors": null
}
```

### Steps To Reproduce

1. Start with a project with more than one dbt model
1. Run `dbt docs generate --select nothing` (where there is no node named `nothing`)
1. Examine `target/catalog.json`
1. Notice that it includes **all** nodes rather than **0** nodes

### Relevant log output

```shell
dbt docs generate --select nothing
```

```
[0m08:42:03.885769 [warn ] [MainThread]: The selection criterion 'nothing' does not match any nodes
[0m08:42:03.886607 [info ] [MainThread]: 
[0m08:42:03.887076 [warn ] [MainThread]: Nothing to do. Try checking your model configs and model specification args
[0m08:42:03.887696 [debug] [MainThread]: Command end result
[0m08:42:03.910804 [debug] [MainThread]: Acquiring new postgres connection 'generate_catalog'
[0m08:42:03.911372 [info ] [MainThread]: Building catalog
[0m08:42:03.913022 [debug] [ThreadPool]: Acquiring new postgres connection 'postgres.information_schema'
[0m08:42:03.924920 [debug] [ThreadPool]: Using postgres connection "postgres.information_schema"
[0m08:42:03.925472 [debug] [ThreadPool]: On postgres.information_schema: BEGIN
[0m08:42:03.925910 [debug] [ThreadPool]: Opening a new connection, currently in state init
[0m08:42:04.114544 [debug] [ThreadPool]: SQL status: BEGIN in 0.0 seconds
[0m08:42:04.115114 [debug] [ThreadPool]: Using postgres connection "postgres.information_schema"
[0m08:42:04.115666 [debug] [ThreadPool]: On postgres.information_schema: /* {"app": "dbt", "dbt_version": "1.7.0b2", "profile_name": "sandcastle", "target_name": "postgres", "connection_name": "postgres.information_schema"} */
```

### Environment

```markdown
- OS: N/A
- Python: N/A
- dbt: `main` branch for 1.7.0-b2
```

### Which database adapter are you using with dbt?

postgres

### Additional Context

In case it is helpful in some way, my hack [here](https://github.com/dbt-labs/dbt-core/issues/8452#issuecomment-1709129455) contains logic for selecting 0 nodes rather than all nodes.

**Comment by graciegoheen** [comment]
We need to add an additional if test for when there are no selected nodes

**Comment by peterallenwebb** [comment]
This behavior was fixed as a secondary benefit of my work on [#8846](https://github.com/dbt-labs/dbt-core/issues/8846). A test demonstrating the fix [can be seen here](https://github.com/dbt-labs/dbt-core/blob/1.7.latest/tests/functional/docs/test_generate.py#L32). The test shows that a selector that doesn't match any nodes generates a catalog with no entries, rather than a catalog with every entry.
